### ISSUE TITLE
Cannot refresh access token without explicit scope

### ISSUE THREAD

**Issue by markhobson** [spring-projects/spring-security-oauth#1416]
The OAuth2 specification states that providing a scope is optional when [refreshing an access token](https://tools.ietf.org/html/rfc6749#section-6):

```
scope
      OPTIONAL.  The scope of the access request as described by
      Section 3.3.  The requested scope MUST NOT include any scope
      not originally granted by the resource owner, and if omitted is
      treated as equal to the scope originally granted by the
      resource owner.
```

Attempting to do this against a Spring Security OAuth2 authorization server results in `DefaultOAuth2RequestValidator` throwing the following exception:

```
InvalidScopeException: Empty scope (either the client or the user is not allowed the requested scopes)
```

See this test to reproduce the issue: https://github.com/markhobson/spring-security-oauth-issue/

**Comment by markhobson** [comment]
Still an issue with spring-security-oauth2 2.3.5. Is there any appetite to fix this or is all effort going into Spring Security 5 OAuth?

**Comment by jgrandja** [comment]
@markhobson Thanks for the report. This is a bug as you have determined. To workaround this issue you can provide a custom `OAuth2RequestValidator` via `AuthorizationServerEndpointsConfigurer.requestValidator()`.

We are quite busy working on features for the upcoming Spring Security 5.2.0 release so yes all our efforts are focused there. I've scheduled this fix for next release but that won't be for a while given we recently released 2.3.5.

Would you be interested in submitting a PR for this fix?

**Comment by markhobson** [comment]
Hi @jgrandja, thanks for the reply. I'm happy to take a look at creating a PR, although I thought it was more involved than updating the `OAuth2RequestValidator` as we need to obtain the originally granted scopes when absent.

**Comment by jgrandja** [comment]
@markhobson 

> we need to obtain the originally granted scopes when absent

Yes, this is required as part of the fix. If you look at `AuthorizationEndpoint`, you will notice it stores `AuthorizationRequest` in the session using the attribute `AuthorizationEndpoint.AUTHORIZATION_REQUEST_ATTR_NAME`.

**Comment by markhobson** [comment]
@jgrandja Your initial workaround of providing a more lenient `OAuth2RequestValidator` was indeed sufficient. Allowing empty scopes results in `DefaultTokenServices` creating an `OAuth2Authentication` with the originally granted scopes. Note that I raised [spring-boot#17014](https://github.com/spring-projects/spring-boot/pull/17014) to ease configuring a custom `OAuth2RequestValidator`.

If I understand correctly, your comment regarding `AuthorizationEndpoint` storing the `AuthorizationRequest` in the session shouldn't be relevant since refreshing is performed by the `TokenEndpoint`? They do both use the `OAuth2RequestValidator` so I trust that changing its behaviour for token refresh won't inadvertently affect authorization.

I'll create a PR so you can review my fix so far.

**Comment by markhobson** [comment]
PR: https://github.com/spring-projects/spring-security-oauth/pull/1697

**Comment by markhobson** [comment]
On reflection I think this PR needs some more work.

Both authorization requests and token requests are passed through the `OAuth2RequestValidator`. At these points requests with empty scopes have been set to the client's scopes. Thus the only case for the validator to allow empty scopes is for a refresh request when the client has no scopes defined. Authorization requests must always resolve to some scope. I'll update the validator to only guard against empty scopes in `validateScope(AuthorizationRequest, ClientDetails)`.

It is a little misleading that for refresh requests empty scopes are translated to client scopes prior to validation, and then reset back to the requested scopes afterwards. This doesn't give the validator a clear view of the actual request.

I believe this whole issue can be worked around if the client is registered with scopes. In my case it wasn't, so coupled with the empty scope refresh request it highlighted this edge case.

**Comment by markhobson** [comment]
Updated PR to only relax token request validation. @jgrandja Let me know your thoughts, thanks.
